java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/db/DBAddConnection.java

```diff
@@ -66,12 +70,66 @@
 @ServiceProvider(service = CodeActionsProvider.class)
 public class DBAddConnection extends CodeActionsProvider {
     public static final String DB_ADD_CONNECTION =  "db.add.connection"; // NOI18N
+    public static final String USER_ID =  "userId"; // NOI18N
+    public static final String PASSWORD =  "password"; // NOI18N
+    public static final String DRIVER =  "driver"; // NOI18N
+    public static final String DB_URL =  "url"; // NOI18N
+    public static final String SCHEMA =  "schema"; // NOI18N
+    public static final String DISPLAY_NAME =  "displayName"; // NOI18N
+    
+    private static final Set<String> COMMANDS = new HashSet<>(Arrays.asList(DB_ADD_CONNECTION));
+    
+    private final Gson gson = new Gson();
 
     @Override
     public CompletableFuture<Object> processCommand(NbCodeLanguageClient client, String command, List<Object> arguments) {
         if (!DB_ADD_CONNECTION.equals(command)) {
             return null;
         }
+        
+        String userId = null;
+        String dbUrl = null;
+        String driverClass = null;
+        
+        final Map m = gson.fromJson(gson.toJson(arguments.get(0)), Map.class);
+        if (m != null) {
+            userId = (String) m.get(USER_ID);
+            dbUrl = (String) m.get(DB_URL);
+            driverClass = (String) m.get(DRIVER);
+            
+        }
+        if (dbUrl != null && driverClass != null) {
+            
+            JDBCDriver[] driver = JDBCDriverManager.getDefault().getDrivers(driverClass); //NOI18N
+            if (driver != null && driver.length > 0) {
+                CompletableFuture<String> usernameFuture = userId != null ? CompletableFuture.completedFuture(userId) : client.showInputBox(new ShowInputBoxParams(
+                        Bundle.MSG_EnterUsername(), userId));
+                
+                usernameFuture.thenAccept((username) -> { //NOI18N
+                    if (username == null) {
+                        return;
+                    }
+                    char[] password = m.get(PASSWORD) == null ? null 
+                            : ((List<Double>) m.get(PASSWORD)).stream().map(n -> Character.toString((char) n.byteValue())).collect(Collectors.joining()).toCharArray();
```

Comment:
This is wrong. The common reason to collect passwords into char arrays is that they can be cleared on the heap, while strings might be interned and thus retained indefinitetely. However at this point the data already went through several transformations (including a string variant) and thus the argument is moot.

This also breaks non-ascii passwords. There are multiple problematic assumptions:

- the sending side must know, that the password is expected to be a sequence of UTF-16 code units (you go through Character)
- the value seems transferred as a JSON number array and the each element is treated as a byte. Once you get out of the range of a single byte the value is truncated and thus wrong.

Reply:
This is still broken. You just rewrote the stream to a normal iteration. You did not fix the core problem. Again: WHY do you insist on some number array? Without any pain you force yourself to have encoding issues, without any benefits.

Reply:
Ah, you didn't notice that the stream was transforming via String, I thought it was part of you complain. Javascript doesn't know array of primitives as java does. It is sending array of `number` which is transformed to a `List<Double>` in the protocol. I don't see any reason to have encoding issues.

Reply:
The still unresolved question is: What is the client sending here? How is the array of `number` generated? Storing a password in an array of `number` will need a definition how the password must be encoded, so where is the that defined? I still claim, that you destroy all characters outside the first 255 codepoints.

Reply:
Unfortunately I cannot check the code where the array of `numbers` is created. I tried to check what happens when I set a password containing characters outside the first 255. I was not able to establish the connection from NetBeans or DB Admin console. So apparently UTF characters are badly handled even before the password gets to this command. I'm switching password to `String` and will try to handle it before the command is called.